kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/DecryptionDekCache.java

```diff
@@ -0,0 +1,126 @@
+import com.github.benmanes.caffeine.cache.RemovalCause;
+
+import io.kroxylicious.filter.encryption.FilterThreadExecutor;
+import io.kroxylicious.filter.encryption.dek.CipherSpec;
+import io.kroxylicious.filter.encryption.dek.Dek;
+import io.kroxylicious.filter.encryption.dek.DekManager;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+/**
+ * A cache of DEKs used on the decryption path.
+ * @param <K> The type of KEK id.
+ * @param <E> The type of encrypted DEK.
+ */
+public class DecryptionDekCache<K, E> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DecryptionDekCache.class);
+
+    public static final int NO_MAX_CACHE_SIZE = -1;
+    private final DekManager<K, E> dekManager;
+
+    record CacheKey<E>(
+                       @Nullable CipherSpec cipherSpec,
+                       @Nullable E edek) {
+        public CacheKey {
+            if (cipherSpec == null ^ edek == null) {
+                throw new IllegalArgumentException();
+            }
+        }
+
+        @SuppressWarnings({ "rawtypes", "unchecked" })
+        private static final CacheKey NONE = new CacheKey(null, null);
+
+        @SuppressWarnings("unchecked")
+        static <E> CacheKey<E> none() {
+            return NONE;
+        }
+
+        public boolean isNone() {
+            return cipherSpec == null || edek == null;
+        }
+    }
+
+    private final AsyncLoadingCache<CacheKey<E>, Dek<E>> decryptorCache;
+
+    public DecryptionDekCache(@NonNull DekManager<K, E> dekManager,
+                              @Nullable Executor dekCacheExecutor,
+                              int dekCacheMaxItems) {
+        this.dekManager = Objects.requireNonNull(dekManager);
+        Caffeine<Object, Object> cache = Caffeine.<CacheKey<E>, Dek<E>> newBuilder();
+        if (dekCacheMaxItems != NO_MAX_CACHE_SIZE) {
+            cache.maximumSize(dekCacheMaxItems);
+        }
+        if (dekCacheExecutor != null) {
+            cache.executor(dekCacheExecutor);
+        }
+        this.decryptorCache = cache
+                .removalListener(this::afterCacheEviction)
+                .buildAsync(this::loadDek);
+    }
+
+    /**
+     * Invoked by Caffeine when a DEK needs to be loaded.
+     * This method is executed on the {@code dekCacheExecutor} passed to the constructor.
+     * @param cacheKey The cache key
+     * @param executor The executor
+     * @return A future
+     */
+    private CompletableFuture<Dek<E>> loadDek(CacheKey<E> cacheKey, Executor executor) {
+        if (cacheKey == null || cacheKey.isNone()) {
+            return CompletableFuture.completedFuture(null);
+        }
+        // This assert is just to appease sonar because it doesn't grok that these things cannot be null due to
+        // the cacheKey.isNone() check above
+        assert cacheKey.edek() != null && cacheKey.cipherSpec() != null;
+        return dekManager.decryptEdek(cacheKey.edek(), cacheKey.cipherSpec())
```

Comment:
should we `destroyForEncrypt()`  the DEK immediately here?

Reply:
A DEK with 0 encryptions is already closed for encrypt.

Reply:
But perhaps we need a test for this.